### PR TITLE
log slow executions of posted tasks

### LINKED ISSUE #1880: metrics: add metrics that tracks asio queues
Right now we use 2 queues: the main IOService queue and the worker queue.
We should track the delay in a metric between when a `post` call happens and the actual execution of the callback.

This is especially important to detect starvation or creation of a backlog in the respective subsystem. One of the theories behind #1764 is that too many merges are happening with too few threads for example.

### PR DESCRIPTION
**Pull request by vogel**
Signed-off-by: Rafał Malinowski <rafal.przemyslaw.malinowski@gmail.com>

# Fixes #1880 

# Description

Allow writing to log file information about tasks created with `Application::postOnMainThread`, `Application::postOnMainThreadWithDelay` and `Application::postOnBackgroundThread` that are taking more than one second to start execution.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [x] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/main/ApplicationImpl.cpp [hunk 1]
```diff
@@ -69,6 +70,12 @@ ApplicationImpl::ApplicationImpl(VirtualClock& clock, Config const& cfg)
     , mStoppingTimer(*this)
     , mMetrics(std::make_unique<medida::MetricsRegistry>())
     , mAppStateCurrent(mMetrics->NewCounter({"app", "state", "current"}))
+    , mPostOnMainThreadDelay(
```
Comment (MonsieurNicolas) [inline src/main/ApplicationImpl.cpp]: metrics.md need to be updated

**Comment by vogel** [comment]
metrics updated

File: src/util/LogSlowExecution.h [hunk 2]
```diff
@@ -9,26 +9,46 @@
...
-    std::chrono::system_clock::time_point mStart;
-    std::string mName;
-    int mThresholdInMs;
-
   public:
-    LogSlowExecution(std::string eventName, int ms = 1000)
+    enum class Mode
+    {
+        EXECUTION,
+        START
```
Comment (marta-lokhova) [inline src/util/LogSlowExecution.h]: I think this enum is confusing: `LogSlowExecution` does not know that time before start will be measured, and `checkExecuted` can be called anywhere. I think it should be more generic, something like this:
```
enum class Mode 
{
    AUTOMATIC_RAII,
    MANUAL // In this mode, it is the caller's responsibility to check elapsed time
}
```

**Comment by vogel** [comment]
Updated and rebased

**Comment by marta-lokhova** [comment]
r+ a15de86d70f284964f22fb4d46422d94fa7088ff

**Pull request merged**